h2o-algos/src/main/java/hex/rulefit/RuleFit.java

```diff
@@ -280,23 +283,29 @@ void fillModelMetrics(RuleFitModel model, GLMModel glmModel) {
             return array;
         }
 
-        SharedTreeModel trainTreeModel(RuleFitModel.Algorithm algorithm, int maxDepth) {
-            SharedTreeModel treeModel;
-            treeParameters._max_depth = maxDepth;
-
-            if (algorithm.equals(RuleFitModel.Algorithm.DRF)) {
-                DRF job = new DRF((DRFModel.DRFParameters) treeParameters);
-                treeModel = job.trainModel().get();
-
-            } else if (algorithm.equals(RuleFitModel.Algorithm.GBM)) {
-                GBM job = new GBM((GBMModel.GBMParameters) treeParameters);
-                treeModel = job.trainModel().get();
-            } else {
-                // TODO XGB
-                throw new RuntimeException("Unsupported algorithm for tree building: " + _parms._algorithm);
+        SharedTree[] getBuilders(RuleFitModel.Algorithm algorithm, int[] depths) {
+            SharedTree[] builders = new SharedTree[depths.length];
+
+            for (int modelId = 0; modelId < depths.length; modelId++) {
+                SharedTreeModel.SharedTreeParameters params = (SharedTreeModel.SharedTreeParameters) treeParameters.clone();
+                params._max_depth = depths[modelId];
+               
+                if (algorithm.equals(RuleFitModel.Algorithm.DRF)) {
+                    builders[modelId] = new DRF((DRFModel.DRFParameters) params);
+                    ((SharedTreeModel.SharedTreeParameters) builders[modelId]._parms)._max_depth = depths[modelId];
+                    DKV.put(builders[modelId]._job);
+                    Scope.track_generic(builders[modelId]._job);
+                } else if (algorithm.equals(RuleFitModel.Algorithm.GBM)) {
+                    builders[modelId] = new GBM((GBMModel.GBMParameters) params);
+                    ((SharedTreeModel.SharedTreeParameters) builders[modelId]._parms)._max_depth = depths[modelId];
+                    DKV.put(builders[modelId]._job);
+                    Scope.track_generic(builders[modelId]._job);
+                } else {
+                    // TODO XGB
```

Comment:
why is there XGB todo in rulefit?

Reply:
because it is planned to use also XGB as algo for rule generation for rulefit but it is not yet done. this TODO was in trainTreeModel() which now won't exist so it's now here in respective place.